api/types/swarm/task.go

```diff
@@ -71,8 +72,30 @@ type TaskSpec struct {
 
 // Resources represents resources (CPU/Memory).
 type Resources struct {
-	NanoCPUs    int64 `json:",omitempty"`
-	MemoryBytes int64 `json:",omitempty"`
+	NanoCPUs         int64              `json:",omitempty"`
+	MemoryBytes      int64              `json:",omitempty"`
+	GenericResources []*GenericResource `json:",omitempty"`
+}
+
+// GenericResource represents a "user defined" resource which can
+// be either an integer (e.g: SSD=3) or a string (e.g: SSD=sda1)
+type GenericResource struct {
+	Str      *GenericString   `json:",omitempty"`
+	Discrete *GenericDiscrete `json:",omitempty"`
+}
+
+// GenericString represents a "user defined" resource which is defined
+// as a string
+type GenericString struct {
+	Kind  string `json:",omitempty"`
```

Comment:
Hi !

Kind is used to describe the Kind of the resource (e.g: GPU, FPGA, ...)
It is then used when creating a service ($ docker service create --name nginx --image nginx:latest --generic-resources "GPU=2")


You can find the design doc here: https://github.com/docker/swarmkit/blob/master/design/generic_resources.md